src/clang-c-frontend/clang_c_convert.cpp

```diff
@@ -1444,7 +1447,50 @@ bool clang_c_convertert::get_expr(const clang::Stmt &stmt, exprt &new_expr)
     if(get_expr(*compound.getInitializer(), initializer))
       return true;
 
-    new_expr = initializer;
+    typet t = initializer.type();
+
+    /* A compound literals is an LValue that has associated static or automatic
+     * storage, depending on whether it appears in file or block scope.
+     * Therefore, in C, for instance a pointer to it can be taken and used
+     * within the same block without restrictions, e.g.
+     *
+     *   int *p = &(int){0}, x = 42;
+     *   memcpy(p, &x, sizeof(int));
+     *   assert(*p == 42);
+     *
+     * It has the same semantics as an non-anonymous variable of the
+     * corresponding type. Thus, we introduce a new symbol to represent it.
+     */
+
+    /* "cl" stands for "compound literal", but that is too long for possibly
+     * frequent copies of the symbol's name and id. */
+    symbolt &cl = anon_symbol.new_symbol(context, initializer.type(), "cl$");
+    /* mimic get_default_symbol() w/o setting id and name as that's already
+     * taken care of by anon_symbol.new_symbol() above */
+    cl.mode = "C";
+    cl.module = get_modulename_from_path(location.file().as_string());
+    cl.location = location;
+    cl.type = t;
```

Comment:
here you're probably want to use `get_decl_name` + `get_default_symbol` to generate the symbol.

You'll probably need to extend `get_decl_name` for compound literals, similar to how it's done for unnamed fields.

Reply:
I've tried to, but failed, since there does not seem to be any `clang::NamedDecl` (or any `clang::Decl`) available at this point.

I've noticed several assumptions being made on the `id` string elsewhere; is it's format defined in ESBMC? For instance, the `c:main.c@54@F@main@z` id above is purely generated by clang, yet there are many instances in which the id is compared to specific constants (mostly for intrinsics and functions in c2goto/library).

Reply:
> Not really, we use it as defined by clang.

Since there is no `clang::Decl` for this anonymous object, I don't see a way to use Clang to generate such an id for us in a meaningful way. The situation for unnamed field members is different in that they are already part of a declaration whereas the anonymous symbol we introduce does not have declaration in the AST.

The comment above the code you cited mentions `get_default_symbol()`. I'm not calling this function, because that would require some awkward re-conversion of `name`/`id` to `std::string` and back to `irep_idt` not changing any of the actual values.

Reply:
> The comment above the code you cited mentions `get_default_symbol()`. I'm not calling this function, because that would require some awkward re-conversion of `name`/`id` to `std::string` and back to `irep_idt` not changing any of the actual values.

Do you mean inside the new class you created? My concern is the duplicated code, as this is basically the code inside `get_default_symbol()`: the goal of this function is to create all symbols in the frontend, so we don't introduce inconsistencies by duplicating the code in other places.

Reply:
> Do you mean inside the new class you created?

I meant this comment directly above: https://github.com/esbmc/esbmc/pull/711/files#diff-77a14a166a14947af9a9605e058ed20d935cf5200891234187cf015166925590R1470-R1471

> My concern is the duplicated code [...]

Right. I've changed the signature of get_default_symbol() to take `irep_idt`s directly, thus avoiding the awkward conversions. The code in question now uses this function.